### DISCUSSION TITLE
😭  How do you specify final state?  The front page docs do not say

### DISCUSSION THREAD

**Discussion by enjoysmath** [pytransitions/transitions#656] [Q&A] [answered]
I'll take a guess that it goes in the same place as 'initial', but who knows ! 🙄

**Comment by enjoysmath** [comment]
I just want to say this is an amazing library.  I didn't get my code working using transitions alone, but I used transitions-gui, which even simulates the machine!  This is freaking amaze balls 🎱

**Reply by aleneum** [reply, marked as answer]
Transitions does not feature a specific "final" state argument. What would you like to achieve with final states? You could wrap up things on "def on_enter_FinalState" or pass a "done" callback to transitions that enter that state.

Furthermore, you could have a look at [state feature decorations](https://github.com/pytransitions/transitions?tab=readme-ov-file#adding-features-to-states). You could add "Tags" or "Error". The later will raise an Exception when a state was entered that cannot be left.

**Reply by aleneum** [reply]
This test case might be useful:

```python
def test_error(self):

        @add_state_features(Error)
        class CustomMachine(Machine):
            pass

        states = ['A', 'B', 'F',
                  {'name': 'S1', 'tags': ['accepted']},
                  {'name': 'S2', 'accepted': True}]

        transitions = [['to_B', ['S1', 'S2'], 'B'], ['go', 'A', 'B'], ['fail', 'B', 'F'],
                       ['success1', 'B', 'S2'], ['success2', 'B', 'S2']]
        m = CustomMachine(states=states, transitions=transitions, auto_transitions=False, initial='A')
        m.go()
        m.success1()
        self.assertTrue(m.get_state(m.state).is_accepted)
        m.to_B()
        m.success2()
        self.assertTrue(m.get_state(m.state).is_accepted)
        m.to_B()
        with self.assertRaises(MachineError):
            m.fail()
```

**Comment by enjoysmath** [comment]
Thanks Alexander.  I need everything to work with Transitions Gui.  I have
a nested state machine:

[image: image.png]

I need to eventually come out of a state in the Slab Heating nested
machine, and back to outer Main Loop's Start, but simply specifying
transition doesn't work.

**Reply by aleneum** [reply]
Unfortunately, the image is not visible. Maybe you could provide a "minimal, reproducible (code) example" (MRE).

**Comment by enjoysmath** [comment]
Alexander,

I got what you were saying about the event . Ah... got to reboot can't type

**Comment by enjoysmath** [comment]
Dr. Neumann,

Your first suggestion worked.  I can exit the nested machine by simply
calling its my_finalize_event.

Thank you for your responsiveness!  Very greatly appreciated and the
library is awesome.  The standard library for this now and hope to be using
it a lot.

Daniel
